ion-schema/src/main/kotlin/com/amazon/ionschema/model/Constraint.kt

```diff
@@ -0,0 +1,236 @@
+package com.amazon.ionschema.model
+
+import com.amazon.ion.IonValue
+import com.amazon.ionschema.IonSchemaVersion
+import com.amazon.ionschema.internal.util.validateRegexPattern
+import kotlin.text.Regex as KRegex
+
+/**
+ * This internal only interface allows us to use an exhaustive `when` within this library.
+ * It will become `sealed` when we update to kotlin language version = 1.5 or higher.
+ *
+ * This is not visible to users of the library. Users of the library must use [Constraint], which is not sealed so that
+ * we can add new constraints (in new versions of Ion Schema) to this library in a way that is backwards compatible for
+ * consumers.
+ *
+ * TODO—some library users might prefer for their builds to break when they are not handling a new constraint. Should
+ * we make this public with the caveat that it requires opt in, and is subject to breaking changes?
```

Comment:
I think we do not want to allow people to implement their own (for now). There was discussion about this as part of the ISL 2.0 spec work. We declined to introduce custom constraints (for now) because there was no concrete use case with which to justify the cost in complexity.

However, one thing we did decide is that _if_ we do add user-defined constraints, we need to have support for it in the Ion Schema Specification, or else we'll start running into problems with schemas not being portable (i.e. different behavior for the same inputs on differently configured systems). For example, a custom constraint using the "is_prime" field name—if Ion Schema Kotlin on one system has that constraint installed, and on another system, Ion Schema Kotlin does not have that constraint installed, then the same schema and the same data could have different validation results on two different machines.